analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs

```diff
@@ -34,19 +34,23 @@ protected override ProgramState LearnBranchingConstraint(ProgramState state, IBi
     {
         if (operation.OperatorKind.IsAnyEquality())
         {
-            state = LearnBranchingConstraint<ObjectConstraint>(state, operation, falseBranch) ?? state;
-            state = LearnBranchingConstraint<BoolConstraint>(state, operation, falseBranch) ?? state;
+            state = LearnBranchingEqualityConstraint<ObjectConstraint>(state, operation, falseBranch) ?? state;
+            state = LearnBranchingEqualityConstraint<BoolConstraint>(state, operation, falseBranch) ?? state;
+        }
+        else if (operation.OperatorKind.IsAnyRelational())
+        {
+            state = LearnBranchingRelationalConstraint(state, operation, falseBranch) ?? state;
         }
         return state;
     }
 
-    private static ProgramState LearnBranchingConstraint<T>(ProgramState state, IBinaryOperationWrapper binary, bool falseBranch)
+    private static ProgramState LearnBranchingEqualityConstraint<T>(ProgramState state, IBinaryOperationWrapper binary, bool falseBranch)
         where T : SymbolicConstraint
     {
         var useOpposite = falseBranch ^ binary.OperatorKind.IsNotEquals();
-        // We can fall through ?? because "constraint" and "testedSymbol" are exclusive. Symbols with the constraint will be recognized as "constraint" side.
-        if ((OperandConstraint(binary.LeftOperand) ?? OperandConstraint(binary.RightOperand)) is { } constraint
-            && (OperandSymbolWithoutConstraint(binary.LeftOperand) ?? OperandSymbolWithoutConstraint(binary.RightOperand)) is { } testedSymbol
+        // We can take the 1st constraint and "testedSymbol" because they are exclusive. Symbols with T constraint will be recognized as the constraining side.
+        if (FirstBinaryOperandConstraint<T>(state, binary) is { } constraint
```

Comment:
I don't like the `First`, because it's not doing `first of N` operation, it's doing "one or other". It works without the `First` too
```suggestion
        if (BinaryOperandConstraint<T>(state, binary) is { } constraint
```